src/Core/Batch/BatchDaemon.php

```diff
@@ -0,0 +1,203 @@
+    private $command;
+
+    /**
+     * Prepare the descriptor spec and install signal handlers.
+     *
+     * @param string $entrypoint Daemon's entrypoint script.
+     */
+    public function __construct($entrypoint)
+    {
+        if (! $this->isSysvIPCLoaded()) {
+            throw new \RuntimeException('SystemV IPC exntensions are missing.');
+        }
+        $this->runner = new BatchRunner(
+            new SysvConfigStorage(),
+            new SysvSubmitter()
+        );
+        $this->shutdown = false;
+        // Just share the usual descriptors.
+        $this->descriptorSpec = [
+            0 => ['file', 'php://stdin', 'r'],
+            1 => ['file', 'php://stdout', 'w'],
+            2 => ['file', 'php://stderr', 'w']
+        ];
+        // setup signal handlers
+        pcntl_signal(SIGTERM, [$this, "sigHandler"]);
+        pcntl_signal(SIGINT, [$this, "sigHandler"]);
+        pcntl_signal(SIGHUP, [$this, "sigHandler"]);
+        pcntl_signal(SIGALRM, [$this, "sigHandler"]);
+        $this->command = sprintf('exec php -d auto_prepend_file="" %s', $entrypoint);
+        $this->initFailureFile();
+    }
+
+    /**
+     * A signal handler for setting the terminate switch.
+     * {@see http://php.net/manual/en/function.pcntl-signal.php}
+     *
+     * @param int $signo The received signal.
+     * @param mixed $siginfo [optional] An array representing the signal
+     *              information. **Defaults to** null.
+     *
+     * @return void
+     */
+    public function sigHandler($signo, $signinfo = null)
+    {
+        switch ($signo) {
+            case SIGINT:
+            case SIGTERM:
+                $this->shutdown = true;
+                break;
+        }
+    }
+
+    /**
+     * A loop for the parent.
+     *
+     * @return void
+     */
+    public function runParent()
+    {
+        $procs = [];
+        while (true) {
+            $jobs = $this->runner->getJobs();
+            foreach ($jobs as $job) {
+                if (! array_key_exists($job->getIdentifier(), $procs)) {
+                    echo 'Spawning children' . PHP_EOL;
+                    $procs[$job->getIdentifier()] = [];
+                    for ($i = 0; $i < $job->getWorkerNum(); $i++) {
+                        $procs[$job->getIdentifier()][] = proc_open(
+                            sprintf('%s %d', $this->command, $job->getIdNum()),
+                            $this->descriptorSpec,
+                            $pipes
+                        );
+                    }
+                }
+            }
+            pcntl_signal_dispatch();
+            if ($this->shutdown) {
+                echo 'Shutting down, waiting for the children' . PHP_EOL;
+                foreach ($procs as $k => $v) {
+                    foreach ($v as $proc) {
+                        $status = proc_get_status($proc);
+                        // Keep sending SIGTERM until the child exits.
+                        while ($status['running'] === true) {
```

Comment:
The child is using `SIGALRM` for timing out the blocking `msg_receive` call. I sometimes observed a single proc_terminate call is dropped. I think this happens when the `SIGTERM` arrives during the signal handler for the `SIGALRM`. Calling it multiple times make sure the `SIGTERM` will arrive (theoretically it can conflict again, but very unlikely).